Emscripten: Screenshot and Logfile support and minor bugfixes & improvements

chkuendig opened the pull request:
I wrapped up a few minor improvements over the holidays. 

**Logfile Export**
Logfiles can now be opened (as a download in the browser), they weren't accessible previously.

https://github.com/scummvm/scummvm/assets/748698/edd8de1a-8f48-40cd-a968-d9181f006f94


**Screenshot support** 


Screenshots can now also be exported. I had to hack this a bit as WebGL clears the drawing buffer after compositing the contents, so we need to wait for the next frame to take the screenshot. But it works nicely in practice.

https://github.com/scummvm/scummvm/assets/748698/c3ef2050-7361-49c4-9b5b-47bdd3793cb2



**Fixing an OpenGL performance regression**
The new debugging code increased frame times 10x, I think this is because there's quite a a lot of warnings/errors when running shaders in WebGL. For now I just disable the debugging code.
<img width="1153" alt="image" src="https://github.com/scummvm/scummvm/assets/748698/eb6145c4-4c2e-4230-bb9c-9b4805beb62b">

**Engines not compiling with OpenGL ES**
`HPL1` and `Freescape` include OpenGL symbol that aren't part of GLES2. I flagged HPL1 as requiring opengl classic and ifdef'ed out the critical parts in Freescape. Freescape still doesn't run as the shaders aren't compatible with OpenGL ES (`Could not compile shader freescape_triangle.fragment: ERROR: 0:33: '[]' : Index expression must be constant.`), but at least it compiles. If there were a good way to force a single engine to TinyGL that would probably be the best fix as performance is good enough either way.  For HPL1 I didn't find any workaround so I think it's okay to exclude it completely from the build.

**Minor cleanups & fixes**
- Adjusted game downloader to the new spreadsheet 
- Disabled GLAD and fixed a build issue related to that - GLAD is not needed on Emscripten as we target Emscripten's emulated OpenGL ES built on top of WebGL and the current GLAD implementation is slow (see Dav1dde/glad-web#12). 
- Properly set up file paths for scummvm.ini
- Set render mode to OpenGL automatically
- Remove gamepad API workaround
- Rewrite build script and some configure logic to be more deterministic and update Emscripten (and remove workarounds)
- Loading game icons directly without having to fetch a large zipfile to simplify things and speed up loading

neuromancer commented:
>  Freescape still doesn't run as the shaders aren't compatible with OpenGL ES (Could not compile shader freescape_triangle.fragment: ERROR: 0:33: '[]' : Index expression must be constant.), but at least it compiles.

This is very useful information, I will fix this ASAP.

neuromancer commented on engines/freescape/gfx_opengl_shaders.cpp:
```diff
@@ -96,7 +96,9 @@ void OpenGLShaderRenderer::init() {
 	_bitmapShader->enableVertexAttribute("position", _bitmapVBO, 2, GL_FLOAT, GL_TRUE, 2 * sizeof(float), 0);
 	_bitmapShader->enableVertexAttribute("texcoord", _bitmapVBO, 2, GL_FLOAT, GL_TRUE, 2 * sizeof(float), 0);
 
+#if !USE_FORCED_GLES2
```
Please revert this change, the use of GL_LIGHTING here is actually not necessary and I will remove it in another commit.

chkuendig replied:
Great, I removed this. Let me know once it also has been removed in master and I'll rebase this branch and rebuild the demo.

neuromancer replied:
Done: https://github.com/scummvm/scummvm/commit/47eb539c3d1381211523d4e1a296c3073e2c3700

mgerhardy commented on backends/graphics/openglsdl/openglsdl-graphics.cpp:
```diff
@@ -437,6 +437,12 @@ bool OpenGLSdlGraphicsManager::loadVideoMode(uint requestedWidth, uint requested
 void OpenGLSdlGraphicsManager::refreshScreen() {
 	// Swap OpenGL buffers
+#ifdef EMSCRIPTEN
+	if(_queuedScreenshot) {
```
minor formatting issue

chkuendig replied:
fixed, thank you!

mgerhardy commented on backends/graphics/openglsdl/openglsdl-graphics.cpp:
```diff
@@ -449,6 +455,11 @@ void OpenGLSdlGraphicsManager::handleResizeImpl(const int width, const int heigh
+void OpenGLSdlGraphicsManager::saveScreenshot() {
+	_queuedScreenshot = true;
+}
+#endif
```
missing newline between functions

chkuendig replied:
fixed, thank you!

mgerhardy commented on backends/graphics3d/openglsdl/openglsdl-graphics3d.cpp:
```diff
@@ -654,6 +654,13 @@ void OpenGLSdlGraphics3dManager::updateScreen() {
 	}
 
+#ifdef EMSCRIPTEN
+	if(_queuedScreenshot) {
```
minor formatting issue

chkuendig replied:
fixed, thank you!

mgerhardy commented on backends/platform/sdl/emscripten/emscripten.cpp:
```diff
@@ -68,4 +89,48 @@ void OSystem_Emscripten::setFeatureState(Feature f, bool enable) {
+Common::Path OSystem_Emscripten::getScreenshotsPath() {
+	return Common::Path("/tmp/");
+}
+Common::Path OSystem_Emscripten::getDefaultIconsPath() {
```
missing newline

chkuendig replied:
fixed, thank you!

mgerhardy commented on configure:
```diff
@@ -3159,47 +3159,65 @@ EOF
 
 		if test "$_debug_build" = yes; then
-			_optimization_level=-O2
+			_optimization_level=-O2 
```
whitespace is not needed here

chkuendig replied:
fixed, thank you!

mgerhardy commented on configure:
```diff
@@ -3159,47 +3159,65 @@ EOF
+			_vorbis=no 
 		fi
-		if test "$_zlib" != no; then
+		if test "$_zlib" = yes; then 
```
trailing whitespace

chkuendig replied:
fixed, thank you! (there was another one also in the check for `$_png` which I also fixed)

mgerhardy commented on configure:
```diff
@@ -3540,6 +3558,7 @@ if test -n "$_host"; then
 			_sdlnet=no
 			_libcurl=no
 			_curl=no
+            _enet=no
```
indention via tab

chkuendig replied:
fixed, thank you!

mgerhardy commented on engines/freescape/gfx_opengl.cpp:
```diff
@@ -441,3 +443,6 @@ Graphics::Surface *OpenGLRenderer::getScreenshot() {
 } // End of namespace Freescape
+
+
+#endif
```
missing newline at the end of the file

mgerhardy commented on graphics/opengl/debug.h:
```diff
@@ -24,7 +24,9 @@
 
 #include "graphics/opengl/context.h"
 
+#ifndef __EMSCRIPTEN__
```
maybe this should get wrapped in in `#ifdef DEBUG` - other platforms might have the same problem.

chkuendig replied:
@ccawley2011 can you chime in? Considering that nobody has noticed this so far since 18 months ( scummvm/scummvm@863988f), I want to be careful to disable it for all platforms, even if it's only in non-debug builds.

ccawley2011 replied:
If it's just slow because there are a lot of warnings, I'd leave this as specific to Emscripten for now.

mgerhardy commented on gui/gui-manager.cpp:
```diff
@@ -112,8 +112,14 @@ void GuiManager::initIconsSet() {
+#ifdef EMSCRIPTEN
+	Common::Path iconsPath = ConfMan.getPath("iconspath");
+	_iconsSet = Common::SearchSet();
+	_iconsSet.addDirectory("gui-icons/",iconsPath,0,3,false);
```
minor formatting issue

chkuendig replied:
fixed, thank you!

neuromancer commented:
Should you be using `defined(USE_OPENGL_GAME)` instead of `USE_OPENGL_GAME`?

Check other examples here: https://github.com/search?q=repo%3Ascummvm%2Fscummvm%20USE_OPENGL_GAME&type=code

Otherwise, freescape is already using `ifdef USE_OPENGL` in their module.mk

ccawley2011 commented on backends/graphics/openglsdl/openglsdl-graphics.cpp:
```diff
@@ -449,6 +455,12 @@ void OpenGLSdlGraphicsManager::handleResizeImpl(const int width, const int heigh
 
+#ifdef EMSCRIPTEN
+void OpenGLSdlGraphicsManager::saveScreenshot() {
+	_queuedScreenshot = true;
```
Would there be a major disadvantage to doing this on all platforms?

chkuendig replied:
It adds a delay of 1 frame and generally seems a bit hackish. I would probably keep this for Emscripten only (I'm not aware of any other platform having this issue)

lephilousophe replied:
Doing this makes sure the screenshot is taken after the whole scene composition is done.
I wonder if we could use a framebuffer as a source for the screenshot instead of taking the next frame.

chkuendig replied:
Unfortunately not all engines support rendering to the framebuffer, so I'd suggest to leave this for now.

ccawley2011 commented on backends/graphics/sdl/sdl-graphics.cpp:
```diff
@@ -363,6 +366,11 @@ void SdlGraphicsManager::saveScreenshot() {
+
+#ifdef EMSCRIPTEN
+		OSystem_Emscripten *emscripten_g_system = dynamic_cast<OSystem_Emscripten*>(g_system);
+		emscripten_g_system->exportFile(screenshotsPath.appendComponent(filename));
```
Could you add a comment for this ifdef?

bluegr replied:
A comment has been added now

ccawley2011 commented on backends/platform/sdl/emscripten/emscripten.cpp:
```diff
@@ -68,4 +89,49 @@ void OSystem_Emscripten::setFeatureState(Feature f, bool enable) {
+	}
+	Common::String exportName = filename.getLastComponent().toString(Common::Path::kNativeSeparator);
+	const int32 size = file.size();
+	char *bytes = new char[size + 1];
```
Does this get deleted after use?

chkuendig replied:
no it doesn't. thanks for the feedback, I'm not used to work in languages without GC...

ccawley2011 commented on graphics/opengl/debug.cpp:
```diff
@@ -39,10 +39,19 @@ Common::String getGLErrStr(GLenum error) {
 		return "GL_INVALID_VALUE";
 	case GL_INVALID_OPERATION:
 		return "GL_INVALID_OPERATION";
+// OpenGL ES 2.0 doesn't have GL_STACK_OVERFLOW and GL_STACK_UNDERFLOW
```
Could you define these in `graphics/opengl/system_headers.h` instead?

chkuendig replied:
The reason I excluded these here is that they aren't returned in GLES 2.0 (see https://registry.khronos.org/OpenGL-Refpages/es2.0/xhtml/glGetError.xml)

But happy to adjust if you think it's clearer to add in.  (or maybe the comment is not precise?)

lephilousophe replied:
You can't use _KHR defines here as it would break the build for platforms not having this define (because it's an extension) and not using GLAD.
I think that what @ccawley2011 proposes is to replicate these defines in `graphics/opengl/system_headers.h` around line 85 to make sure they exist.

chkuendig replied:
Thank you for the clarification, I added them now to the system headers.

ccawley2011 commented on ports.mk:
```diff
@@ -668,6 +668,38 @@ endif
 	devtools/release-checks.sh
 
+# Special generic target for emscripten static file hosting bundle
+dist-emscripten: $(EXECUTABLE) $(PLUGINS)
```
This should be in a port-specific makefile instead.

ccawley2011 commented:
> Otherwise, freescape is already using `ifdef USE_OPENGL` in their module.mk

That checks if OpenGL in general is available, not if the fixed function API is available. For that, you'll want to use `USE_OPENGL_GAME` instead (which should work in makefiles as well as the preprocessor).

chkuendig commented on engines/hpl1/configure.engine:
```diff
@@ -1,3 +1,3 @@
 # This file is included from the main "configure" script
 # add_engine [name] [desc] [build-by-default] [subengines] [base games] [deps]
-add_engine hpl1 "Hpl1" no "" "" "16bit highres jpeg gif png opengl_game_shaders"
+add_engine hpl1 "Hpl1" no "" "" "16bit highres jpeg gif png opengl_game_classic"
```
I'll do that. I guess there's no way of limiting it to "OpenGL but not OpenGL ES" in configure.engine, right?

chkuendig replied:
To follow up on the above, I'm not sure what's the best way to automatically disable an engine that depends on shaders in OpenGL but needs the "Desktop OpenGL" API, not OpenGL ES.

lephilousophe replied:
Engines supporting OpenGL with shaders are expected to support GLES2.
If this is not the case, then it's a problem with the engine: it also means that the engine can't work on Android and iOS.

sev- replied:
Yes, instead of modifying this, probably, you could just disable the engine in your build scripts.

chkuendig replied:
done, removed this commit competely.

chkuendig commented:
Thanks everybody for the thorough review and advise. 

I pushed an update that should address all points where there were no more questions open.

lephilousophe commented:
Thank you for this update.
As I understand this PR, Emscripten is not stuck anymore on an old version.
Could you update the README to reflect this new fact?

Now that the toolchain can use up to date compilers, I will try to find time to dockerize it and add the platform to the buildbot.

chkuendig commented:
I removed that bug from the README. Newest Emscripten is supported and almost more importantly we don't need any custom patches anymore either for any dependency.

lephilousophe commented on backends/graphics/openglsdl/openglsdl-graphics.h:
```diff
@@ -69,6 +69,21 @@ class OpenGLSdlGraphicsManager : public OpenGL::OpenGLGraphicsManager, public Sd
+	 *    Techniques like synchronous drawing buffer access (e.g., calling readPixels or toDataURL in the same function
+	 *    that renders to the drawing buffer) can be used to get the contents of the drawing buffer >>
+	 * 
+	 * This means we need to take the screenshot before drawing, which we do by queueing taking the screenshot
```
This is not taken before but after drawing.

sev- commented on backends/platform/sdl/emscripten/emscripten.mk:
```diff
@@ -0,0 +1,32 @@
+	$(EMSDK_NODE) "$(srcdir)/dists/emscripten/build-make_http_index.js" ./build-emscripten/data
+	cp "$(srcdir)/dists/emscripten/assets/"* ./build-emscripten/
+	cp "$(srcdir)/gui/themes/common-svg/logo.svg" ./build-emscripten/
+	cp "$(srcdir)/icons/scummvm.ico" ./build-emscripten/favicon.ico
```
end line is missing

chkuendig replied:
added now, thanks!

sev- reviewed (approved):
Looks mostly good to me, only the HPL1 engine change has to be reverted, and there is a missing newline in one of the new files.

chkuendig commented:
I removed the HPL commit, added the newline in the makefile, added the OPENGL defines to the system headers and also did a minor edit to the rationale about the screenshot queueing (so it doesn't confuse before/after drawing/compositing anymore, which isn't really that relevant)

lephilousophe commented on dists/emscripten/README.md:
```diff
@@ -69,20 +69,18 @@ In general, ScummVM runs in the browser sufficiently to run all demos and freewa
-*   Don't use asyncify but rewrite main loop to improve performance
-*   Look into emscripten-core/emscripten#16779 as an alternative
+*   Don't use asyncify but rewrite main loop to improve performance.
+*   Look into Stack Switching (emscripten-core/emscripten#16779) or multithreading as an alternative tp Asyncify.
```
There is a small typo here.

chkuendig replied:
ouch. fixed now.

chkuendig commented:
Thanks everybody who reviewed this before. I realized that some of the errors in CI were not flacky tests (none were!), so I pushed again to update this. I think all issues are now resolved.

bluegr commented on dists/emscripten/build.sh:
```diff
@@ -374,8 +367,8 @@ if [[ "run" =~ $(echo ^\(${TASKS}\)$) ]]; then
 
-  EMSDK_NPX=$(dirname $EMSDK_NODE)/npx
-  $EMSDK_NPX -p node-static static .
+  # https://github.com/cloudhead/node-static/issues/241 is menas node-static doesn't work either.
```
```suggestion
  # https://github.com/cloudhead/node-static/issues/241 is means node-static doesn't work either.
```

chkuendig replied:
fixed

bluegr commented on dists/emscripten/build.sh:
```diff
@@ -143,101 +168,91 @@ cd "$ROOT_FOLDER"
-    patch -Np1 -i "$DIST_FOLDER/libmad-0.15.1b-fixes-1.patch"
-    emconfigure ./configure --host=wasm32-unknown-none --build=wasm32-unknown-none --prefix="$LIBS_FOLDER/build/" --with-pic --enable-fpm=no
-    emmake make
+    # libmad needs patching: https://wiki.linuxfromscratch.org/blfs/wiki/libmad
```
The libmad page doesn't exist in the wiki

chkuendig commented:
Its been a week since the last review. Anything else remaining?

sev- commented:
Thank you, merging this. However, it is not clear, what is wrong with the icons path?

chkuendig commented:
> what is wrong with the icons path?

The icons are loaded from a single zipped dat file. On the web this isnt ideal. Adding it as a subfolder to the paths allows us to serve the icons one by one without fetching a larger file. The idea is that all icons are always on the server meaning the dat file can grow very large.
